pravega/pravega
# Issue 5975: ZKSegmentContainerMonitor blocks ForkJoinPool threads when queuing Segment Container Starts

**Change log description**  
Use `AsyncSemaphore` when throttling the number of parallel Segment Container restarts.

**Purpose of the change**  
Fixes #5975.

**What the code does**  
Replaces the existing `Semaphore` in `ZKSegmentContainerMonitor` that controls the number of parallel Segment Container starts by `AsyncSemaphore`. Functionality should be the same, but not blocking calling threads.

**How to verify it**  
Build should pass.
System tests passed:
pravega-development-system-tests-1/4808
pravega-development-system-tests-1/4810

With Pravega Standalone, I can see the new log added and how the parallel Segment Container starts is being respected:
```
2022-10-10 19:32:57,386 3498 [core-16] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Current ongoing Segment Container starts 2.
...
2022-10-10 19:32:57,388 3500 [core-7] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Current ongoing Segment Container starts 2.
...
2022-10-10 19:32:57,424 3536 [core-8] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Current ongoing Segment Container starts 1.
...
2022-10-10 19:32:57,445 3557 [core-4] INFO  i.p.s.s.h.ZKSegmentContainerMonitor - Current ongoing Segment Container starts 0.
```
